docs/layout-of-source-files.rst

```diff
@@ -216,101 +216,562 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
+
+   .. code-block:: bash
+
+       echo 'import "./util.sol"; contract C {}' | solc -
+
+   The content of the standard input is identified in the virtual filesystem by a special import key:
+   ``<stdin>``.
+
+.. warning::
+
+    The compiler uses import keys to determine whether imports refer to the same module or not.
+    If you refer to a file in multiple ways that translate to different keys, it will be compiled
+    multiple times.
+
+    For example:
+
+    .. code-block:: solidity
+        :caption: /code/contracts/contract.sol
+
+        import "tokens/token.sol"   // Import key: tokens/token.sol
+        import "tokens///token.sol" // Import key: tokens///token.sol
+
+    .. code-block:: bash
+
+        cd /code
+        solc tokens/contract.sol /code/tokens/token.sol # Import key: /code/tokens/token.sol
+
+    In the above ``token.sol`` will end up in the virtual filesystem under three different
+    import keys even though all the paths refer to the same file in the underlying filesystem.
+
+    To avoid this situation it is recommended to always use the canonical form of paths in your
+    imports and to only list the top-level files that are not imported by other files when
+    invoking the CLI compiler.
+
+Now that we know how the virtual filesystem works, let us go through the rules used to translate
+import paths into import keys in more detail.
+
+.. index:: absolute import
+
+Absolute Imports
+~~~~~~~~~~~~~~~~
+
+An *absolute import* always starts with a forward slash (``/``).
+The import path translates directly to an import key without normalization of any kind:
+
+::
+
+    import "/project/lib/util.sol" as util;          // Import key: /project/lib/util.sol
+    import "/project/lib/../lib///math.sol" as math; // Import key: /project/lib/../lib///math.sol
+
+In the above you might expect the import key to be reduced to ``/project/lib/math.sol`` but it is
+in fact ``/project/lib/../lib///math.sol``, exactly as stated in the file.
+
+If no file is present under that key in the virtual filesystem, the file loader will also use it as
+is for filesystem lookup.
+The resulting filesystem path is not affected by the value of base path.
+
+.. index:: import relative to base, relative import
+
+.. _imports-relative-to-base:
+
+Imports Relative to Base
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Any import that does not start with ``/``, ``./`` or ``../`` is an *import relative to base*.
+
+::
+
+    import "lib/util.sol" as util;                   // Import key: lib/util.sol
+    import "@openzeppelin/address.sol" as address;   // Import key: @openzeppelin/address.sol
+    import "https://example.com/token.sol" as token; // Import key: https://example.com/token.sol
+
+There is no difference between such imports and absolute ones at the the virtual filesystem level.
+The compiler sees both as opaque identifiers and there is no normalization involved:
+
+::
+
+    import "lib/../lib///math.sol" as math; // Import key: lib/../lib///math.sol
```

Comment:
After writing this all down, I think that not normalizing imports makes behavior more complex than it needs to be and causes tons of unexpected corner cases. Even more so given that we do normalize them in some cases (i.e. in imports relative to source). I think that we should either do some limited normalization on all imports (i.e. remove duplicate slashes, collapse `./` and `../` segments where possible, etc.), or import paths to be already normalized.